refactor for modern features, `new Buffer` deprecation & tooling update

zackschuster opened the pull request:
Hiya!

I'm currently in the process of forking this to run as a serverless function app, mainly by  converting it to an ESM, Rollup-friendly version so I can single-file it up to Azure (Node.js dep management is... not good there). The first thing I did was convert all the code to use ES2015+ features, in the process replacing the deprecated `Buffer` calls with their officially sanctioned alternatives. 

Unsure if the work is wanted upstream, but the tests all pass (on macOS) so I thought I'd at least float it :)

I also removed `iconv` as a dependency since it isn't called anywhere, and won't build on my Windows machine.

Side note: Six tests fail on Windows (`very large text message`, `very large text data`, `html data`, `html file`, `html with image embed` and `html data and attachment`) due to linefeed discrepancies. I didn't change any code to address that, but I thought I'd bring it up. I can go after that in a separate patch if desired.

Thanks & have a nice day! :)

eleith commented:
@zackschuster this is awesome. with such a re-write, it would be helpful to wire this up with eslint/prettier to standardize on style convention?

i would definitely be interested in reviewing this and merging.

zackschuster commented:
sure, i can set up tooling when i have a chance.

i also found out why the tests were failing; the raw fixture inputs weren’t having their line endings normalized (unlike the message output). i can add the fix to this PR as well.

zackschuster commented:
added tooling. also went ahead and updated the tests, since the changes were pretty minor & didn't break anything for macOS.

eleith commented on smtp/client.js:
```diff
@@ -1,209 +1,186 @@
+		}
+		// wait around 1 seconds in case something does come in, otherwise close out SMTP connection if still open
+		else if (this.smtp.state() == smtp.state.CONNECTED)
+			this.timer = setTimeout(() => this.smtp.quit(), 1000);
```
the eslint/prettier settings i use now force brackets around conditional clauses, even if only one line.

other areas in this code do include the brackets, so i'ld like to shoot for at least consistency. 

it seems like you are integrating eslint/prettier with this PR. did prettier run on this file? or is there a rule that would be needed to turn on to get the brackets to be injected?

i think i am used to the airbnb recommendations, so not sure. what do you think?

eleith commented:
awesome.

this looks like mostly a straight forward rewrite that adopts some modern javascript while remaining compatible with node 6, so that is a plus.

i don't have a quick way of verifying it works with windows or mac, so i'll take your word for it =]

the only comment i had was about if we could adopt forcing brackets on conditionals, even for single line statements. not a stickler for it, but worth a conversation to see if it is a quick change on your end.

otherwise, everything looks good, and i'll merge after closing the loop on the above.

thanks again, this provides a much better code base to bug fix now =]

eleith commented:
thanks again, looking forward to merging.

zackschuster commented:
no problem, i’ll make the rule change as soon as i can :)

i can also switch to airbnb style for eslint if preferred.

zackschuster commented:
fixed the braces issue; merely had to turn on `"curly"` with eslint. also added a .editorconfig, since vscode wants to be CRLF on windows :/

here's confirmation the tests run on windows:
![capture](https://user-images.githubusercontent.com/9667758/40930113-4b69edf4-6816-11e8-9bf6-59f08e3d879d.PNG)

i'll post a similar screenshot from my macOS machine.

zackschuster commented:
<img width="665" alt="capture2" src="https://user-images.githubusercontent.com/9667758/40932369-a80b7620-681d-11e8-942d-9c041abeab4f.PNG">

zackschuster commented:
oh realized i should mention that i `.gitignore`'d both `yarn.lock` and `package-lock.json`, so i wouldn't add a lockfile (in case that was an issue). i can also pop either in if you'd like.

eleith commented:
great. i tried to look through all the changes and compare to provide some extra assurance.

i did my best =]

thankfully at least the test case give us some assurances on those usages!

zackschuster commented:
thanks for merging!

i can say that my refactor has been in use in production for some time & hasn’t let us down yet. if i find any bugs however i’ll make sure to push them upstream.

in the meantime, if you’re interested in more work i can package more of the changes from my fork. i’d personally be really interested in getting to the point where i don’t “need” a fork & can use the canonical lib in a serverless setting.

up to you though! thanks again for your attention with my PR :)

eleith commented:
what changes do you have in mind? certainly happy to consider changes that:

(1) can't be done outside the library
(2) are core to the purpose of this module
(3) come with tests

i have a project in mind to add flow to this library to help battle test this a bit more and make moving around pieces a bit easier.

zackschuster commented:
two things, to start with:

1) converting to ESM w/ a single-file bistro
2) as a prerequisite of that, dropping moment.js (since it's massive) for a custom RFC2822 date generator (of which I have some code already [here](https://github.com/zackschuster/smtp/blob/master/lib/message.js#L14)). there'd be tests for that, of course.

I can also look into adding flow, as well as a lockfile (I use yarn, but we can do npm if you prefer)

eleith commented:
i'm definitely open to dropping moment.

what is your goal with (1) ? are you referring to using this-> https://github.com/standard-things/esm

zackschuster commented:
for my purposes, easier serverless deployments: the published `main` file would be a CJS module, ideally with all the deps & such baked in (hence dropping moment).

as a bonus, having ESM source would allow tooling to provide more insight / allow downstream the option to a) compile against source, b) simply require the bundle, or c) integrate with a loader like `esm`.

zackschuster commented:
(also i'm a fan of [this argument](https://medium.com/@Rich_Harris/how-to-not-break-the-internet-with-this-one-weird-trick-e3e2d57fee28) by rich harris)

eleith commented:
what does this look like for a module end user? how can they choose to use a single bundled version of the library versus the unbundled version?

zackschuster commented:
the bundled version would be commonjs, and the api would be exactly the same. this would be the file listed as “main” in package.json.

the source would have the entry file listed as “module” in package.json, with the assumption that the end user would be using tooling of some kind or implementing a loader hook.
